### ISSUE TITLE
jsdom is no longer supported

### ISSUE THREAD

**Issue by ronyeh** [vexflow/vexflow#102]
In my tests while working on PR https://github.com/vexflow/vexflow/pull/98 I have noticed that **jsdom** is no longer supported.

This is because VexFlow 5 requires that the environment supports some sort of font loading. VexFlow 4 just includes all the paths, so no actual WOFF2 or OTF files need to be loaded at runtime.

Even if we bundle the WOFF2 into the vexflow-bravura.js library (as base64), we will still need to "load" the base64 font at runtime so that when the Canvas or SVG context places the SMuFL glyph, it will display correctly.

I fixed a bunch of the old demos under vexflow/demos/, and was able to get [**node-canvas**](https://github.com/Automattic/node-canvas) working after some tweaks. This is because node canvas has a font loading mechanism:

https://github.com/Automattic/node-canvas#registerfont

Anyways, this probably means that the older jsdom backend for our visual regression tests no longer works. But we're mostly using puppeteer these days, so that's not a huge problem.

Anyways, this is something that we should write down in a document "VexFlow 5 vs VexFlow 4" to highlight the differences. If a developer really really needs to use jsdom, they need to use VexFlow 4.

... until we figure out a good solution.

For now, I plan to delete the jsdom demos from the repo. We can always restore them from the V4 repo if needed.

**Comment by ronyeh** [comment]
Actually, jsdom uses node canvas, so we might be able to call node canvas's registerFont() ahead of time and make something work.

I'll investigate.

**Comment by mscuthbert** [comment]
> we should write down in a document "VexFlow 5 vs VexFlow 4" to highlight the differences

That would be very helpful to me.  I've sort of lost track of what the main goals of the v5 release plan to be and that's part of why I haven't been contributing (that plus lack of time).

**Comment by rvilarl** [comment]
Hi @mscuthbert, good to have you back. Main V5 topic is used of OTF SMuFL directly, there is no longer a src/fonts. If you clone it and have a look at tests/flow.html you will see a few new supported fonts!

**Comment by ronyeh** [comment]
I started a document: https://github.com/vexflow/vexflow/wiki/VexFlow-5-vs-VexFlow-4

@rvilarl feel free to add whatever you think the V5 benefits (and drawbacks) are.

Any big change comes with tradeoffs. We hope to (eventually) minimize the drawbacks and make people *want* to upgrade to V5. :-)

**Comment by sschmidTU** [comment]
> this probably means that the older jsdom backend for our visual regression tests no longer works. But we're mostly using puppeteer these days, so that's not a huge problem.

You fixed this in https://github.com/vexflow/vexflow/pull/98#issuecomment-1722452632, right?

Just noting that puppeteer is a lot slower than jsdom/node canvas. That's why OSMD stopped using puppeteer (though it's still an option). Generating all the PNGs for the visual regression tests took about twice as long with puppeteer:
https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/issues/670#issuecomment-592668819

Generating a single png even took about 6 times as long, probably because of the startup time of launching a chromium instance for puppeteer.
puppeteer does have the advantage of being closer to what the browser does (since it uses a browser), so I think there are slight rendering differences, and some methods aren't available without a browser, need to be hacked in, etc. - but for OSMD's visual regression tests and png/svg export scripts, it's great to have a faster option.

**Comment by ronyeh** [comment]
Thanks @sschmidTU 

Good to know that jsdom support is still valuable to some people. It does support custom fonts (through node canvas), but we have to take the extra step of registering the fonts before running the PNG generation. 

But puppeteer is definitely more accurate since it is the real browser behind the scenes. Going forward we will keep jsdom support as long as it is feasible, but our focus will be with puppeteer.

**Comment by sschmidTU** [comment]
Yeah, all good, just wanted to mention it. OSMD doesn't need Vexflow to have jsdom support I think, because OSMD can still use jsdom anyways.
If anyone needs jsdom desperately in Vexflow5, they can still improve it in a PR.
